geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala

```diff
@@ -0,0 +1,90 @@
+/***********************************************************************
+ * Copyright (c) 2013-2019 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.redis.data
+
+import org.geotools.data.Query
+import org.locationtech.geomesa.index.geotools.GeoMesaDataStore
+import org.locationtech.geomesa.index.geotools.GeoMesaDataStoreFactory.GeoMesaDataStoreConfig
+import org.locationtech.geomesa.index.metadata.{GeoMesaMetadata, MetadataStringSerializer}
+import org.locationtech.geomesa.index.stats.GeoMesaStats
+import org.locationtech.geomesa.index.utils._
+import org.locationtech.geomesa.redis.data.RedisDataStore.RedisDataStoreConfig
+import org.locationtech.geomesa.redis.data.index.{RedisIndexAdapter, RedisQueryPlan}
+import org.locationtech.geomesa.redis.data.util.{RedisBackedMetadata, RedisGeoMesaStats, RedisLocking}
+import org.locationtech.geomesa.security.AuthorizationsProvider
+import org.locationtech.geomesa.utils.audit.{AuditProvider, AuditWriter}
+import org.locationtech.geomesa.utils.index.VisibilityLevel
+import org.locationtech.geomesa.utils.io.CloseWithLogging
+import org.opengis.feature.simple.SimpleFeatureType
+import redis.clients.jedis.JedisPool
+
+/**
+  * Data store backed by Redis. Uses Redis SortedSets for range scanning
+  *
+  * @param connection connection pool
+  * @param config datastore configuration
+  */
+class RedisDataStore(val connection: JedisPool, override val config: RedisDataStoreConfig)
+    extends GeoMesaDataStore[RedisDataStore](config) with RedisLocking {
+
+  import org.locationtech.geomesa.utils.geotools.RichAttributeDescriptors.RichAttributeDescriptor
+  import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+
+  import scala.collection.JavaConverters._
+
+  override val metadata: GeoMesaMetadata[String] =
+    new RedisBackedMetadata(connection, config.catalog, MetadataStringSerializer)
+
+  override val adapter: RedisIndexAdapter = new RedisIndexAdapter(this)
+
+  override val stats: GeoMesaStats = RedisGeoMesaStats(this)
+
+  @throws(classOf[IllegalArgumentException])
+  override protected def preSchemaCreate(sft: SimpleFeatureType): Unit = {
+    if (sft.getVisibilityLevel == VisibilityLevel.Attribute) {
+      throw new IllegalArgumentException("Attribute level visibility is not supported in this store")
+    }
+
+    sft.getAttributeDescriptors.asScala.foreach { descriptor =>
+      if (descriptor.getColumnGroups().nonEmpty) {
+        throw new IllegalArgumentException("Column groups are not supported in this store")
```

Comment:
Do all stores throw exceptions like this for unsupported sft options?  

I wonder if it'd be worth taking an approach like ContentDataStore where an implementing class sets various things to true or false to indicate which things are supported/unsupported.

Reply:
that's not a bad idea, although i'd prefer it to be declarative instead of knowing that you have to override something. I opened a ticket here: https://geomesa.atlassian.net/browse/GEOMESA-2565